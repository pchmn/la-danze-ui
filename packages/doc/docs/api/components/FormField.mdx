---
title: FormField
hide_title: true
---

# `FormField`

This component extends `Controller` component of [react-hook-form](https://react-hook-form.com/) and is meant to be used with [material-ui](https://next.material-ui.com/) components.

> It must be used inside a Form component

## Example

```tsx
import { useForm } from 'react-hook-form';
import TextField from '@material-ui/core/TextField';
import { Form, FormField } from 'la-danze-ui';

function App() {
  const form = useForm();

  return (
    <Form form={form}>
      <FormField
        name="email"
        as={TextField}
        rules={{ required: { value: true, message: 'Email is required' } }} />
      ...
    </Form>
  );
}
```

## Props

> Extends [`Controller`](https://react-hook-form.com/api#Controller) props

Additional props:

import {PropsTable} from '/src/components/PropsTable';
export const additionalProps = [
  {
    name: 'error',
    type: 'boolean',
    description: 'Weither field is in error or not.'
  },
  {
    name: 'helperText',
    type: 'string',
    description: 'Error message to display.'
  },
  {
    name: 'disabled',
    type: 'boolean',
    description: 'Weither field is disabled or not.'
  },
];

<PropsTable componentProps={additionalProps} />
