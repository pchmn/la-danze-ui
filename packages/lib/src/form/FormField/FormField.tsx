import React from "react";
import { Controller, ControllerProps, FieldValues } from "react-hook-form";

type NativeInputs = 'input' | 'select' | 'textarea';

type Props = ControllerProps<React.ReactElement | React.ComponentType<any> | NativeInputs, FieldValues> & FormProps;

type FormProps = {
  name: string;
  className?: string;
  label?: string;
  type?: React.InputHTMLAttributes<unknown>['type'];
  error?: boolean;
  helperText?: string;
  disabled?: boolean;
}

export function FormField({ defaultValue = '', ...props }: Props): JSX.Element {
  return (
    <Controller
      defaultValue={defaultValue}
      {...props}
    />
  );

}