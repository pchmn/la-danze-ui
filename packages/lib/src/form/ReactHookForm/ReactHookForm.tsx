import React from "react";
import { UseFormMethods } from "react-hook-form";
import { FormField } from "../FormField/FormField";

interface FormProps extends React.FormHTMLAttributes<any> {
  form: UseFormMethods<any> & { onSubmit?: any };
  disabled?: boolean;
}

export function ReactHookForm({ children, form, disabled, ...otherProps }: FormProps): JSX.Element {

  function childrenWithProps() {
    return React.Children.map(children, child => {
      // Check if child is FormikTextField
      if (React.isValidElement(child) && child.type === FormField) {
        return React.cloneElement(child, {
          control: child?.props?.control || form.control,
          error: child?.props?.error || Boolean(form.errors[child.props.name]?.type),
          helperText: child?.props?.helperText || form.errors[child.props.name]?.message,
          disabled: child?.props?.disabled || disabled
        });
      }
      return child;
    });
  }

  return (
    <form {...otherProps}>
      {childrenWithProps()}
    </form>
  );
}