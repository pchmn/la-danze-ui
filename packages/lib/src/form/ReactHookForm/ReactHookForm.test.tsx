
import TextField from '@material-ui/core/TextField';
import { waitFor } from '@testing-library/react';
import { mount } from 'enzyme';
import React from 'react';
import { useForm, UseFormMethods } from 'react-hook-form';
import { FormField } from '../FormField/FormField';
import { ReactHookForm } from './ReactHookForm';

const onSubmitMock = jest.fn((values: any) => console.log('form values', values));
let useFormValue: UseFormMethods<any>;
describe('<ReactHookForm /> with no props', () => {
  const originalConsoleError = console.error;

  beforeEach(() => {
    console.error = jest.fn((msg) => {
      if (msg.includes('Warning: useLayoutEffect does nothing on the server')) {
        return null;
      } else {
        return originalConsoleError(msg);
      }
    });
  });

  afterEach(() => {
    console.error = originalConsoleError;
  });

  const ContainerWithProps = () => {
    const form = useForm();

    return (
      <ReactHookForm form={form} method="get" action="test" className="className">
        <FormField className="className" name="custom" as={TextField} error={true} type="password" helperText="helper text" disabled label="Password" />
      </ReactHookForm>
    );
  };
  const containerWithProps = mount(<ContainerWithProps />);

  const Container = () => {
    useFormValue = useForm();

    return (
      <ReactHookForm form={useFormValue} onSubmit={useFormValue.handleSubmit(onSubmitMock)}>
        <FormField name="email" as={TextField} rules={{ required: { value: true, message: 'required' }, pattern: { value: /^\S+@\S+$/, message: 'wrong email format' } }} label="Email" />
        <FormField type="password" name="password" as={TextField} rules={{ required: { value: true, message: 'required' }, minLength: { value: 8, message: 'password too short' } }} label="Password" />
        <FormField type="number" name="age" as={TextField} rules={{ required: { value: true, message: 'required' }, min: { value: 18, message: 'you must be 18' }, max: { value: 99, message: 'you are too old' } }} label="Age" />

        <button type="submit">
          Validate
        </button>
      </ReactHookForm>
    );
  };
  const container = mount(<Container />);

  const form = container.find('form');

  it('should match the snapshot', () => {
    expect(container.html()).toMatchSnapshot();
  });

  it('should preserve props of <form> and FormField', () => {
    // Form
    expect(containerWithProps.find('form').first().props()).toEqual(expect.objectContaining({
      method: 'get',
      action: 'test',
      className: 'className'
    }));
    // Formfield
    expect(containerWithProps.find(FormField).first().props()).toEqual(expect.objectContaining({
      name: 'custom',
      className: 'className',
      type: 'password',
      error: true,
      helperText: 'helper text',
      disabled: true,
      as: TextField,
      label: 'Password'
    }));
  });

  it('should add proper props to FormField automatically', () => {
    const formFields = container.find(FormField);

    formFields.forEach(wrapper => {
      expect(wrapper.props()).toEqual(expect.objectContaining({
        control: useFormValue.control,
        error: Boolean(useFormValue?.errors[wrapper.prop('name')]?.type),
        helperText: useFormValue?.errors[wrapper.prop('name')]?.message,
        disabled: undefined
      }));
    });

    // Email
    expect(formFields.first().props()).toEqual(expect.objectContaining({
      name: 'email',
      as: TextField,
      label: 'Email',
      rules: { required: { value: true, message: 'required' }, pattern: { value: /^\S+@\S+$/, message: 'wrong email format' } }
    }));

    // Password
    expect(formFields.at(1).props()).toEqual(expect.objectContaining({
      name: 'password',
      as: TextField,
      label: 'Password',
      rules: { required: { value: true, message: 'required' }, minLength: { value: 8, message: 'password too short' } }
    }));

    // Age
    expect(formFields.at(2).props()).toEqual(expect.objectContaining({
      name: 'age',
      as: TextField,
      label: 'Age',
      rules: { required: { value: true, message: 'required' }, min: { value: 18, message: 'you must be 18' }, max: { value: 99, message: 'you are too old' } }
    }));
  });

  test('it should check mandatory fields', async () => {
    // Simulate submit
    await waitFor(() => form.simulate('submit'))
    container.update();

    // Check errors
    // All fields mandatory
    container.find(FormField).forEach(wrapper => {
      expect(wrapper.props()).toEqual(expect.objectContaining({
        error: true,
        helperText: 'required'
      }));
    });
    expect(onSubmitMock).not.toHaveBeenCalled();
  });

  test('it should check other errors', async () => {
    const inputs = container.find('input');
    inputs.first().simulate('change', {
      target: { value: 'wrongemail' }
    });
    inputs.at(1).simulate('change', {
      target: { value: '1234567' }
    });
    inputs.at(2).simulate('change', {
      target: { value: 17 }
    });
    // Simulate submit
    await waitFor(() => form.simulate('submit'))
    container.update();

    // Check errors
    const formFields = container.find(FormField);
    expect(formFields.first().props()).toEqual(expect.objectContaining({
      error: true,
      helperText: 'wrong email format'
    }));
    expect(formFields.at(1).props()).toEqual(expect.objectContaining({
      error: true,
      helperText: 'password too short'
    }));
    expect(formFields.at(2).props()).toEqual(expect.objectContaining({
      error: true,
      helperText: 'you must be 18'
    }));
    expect(onSubmitMock).not.toHaveBeenCalled();
  });

  test('it should check other errors 2', async () => {
    container.find('input[name="age"]').first().simulate('change', {
      target: { value: 100 }
    });
    // Simulate submit
    await waitFor(() => form.simulate('submit'))
    container.update();

    // Check errors
    expect(container.find(FormField).at(2).props()).toEqual(expect.objectContaining({
      error: true,
      helperText: 'you are too old'
    }));
    expect(onSubmitMock).not.toHaveBeenCalled();
  });

  test('it should call onSubmit function', async () => {
    const inputs = container.find('input');
    inputs.first().simulate('change', {
      target: { value: 'good@email' }
    });
    inputs.at(1).simulate('change', {
      target: { value: '12345678' }
    });
    inputs.at(2).simulate('change', {
      target: { value: 20 }
    });
    // Simulate submit
    await waitFor(() => form.simulate('submit'))
    container.update();

    // Check onSubmit callback
    expect(onSubmitMock).toHaveBeenCalled();
    expect(onSubmitMock.mock.calls[0][0]).toEqual({ email: 'good@email', password: '12345678', age: 20 });
  });
});