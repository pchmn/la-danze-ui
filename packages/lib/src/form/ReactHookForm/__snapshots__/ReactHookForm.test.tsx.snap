// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReactHookForm /> with no props should match the snapshot 1`] = `"<form><div class=\\"MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense MuiInputLabel-outlined\\" data-shrink=\\"false\\">Email</label><div class=\\"MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense\\"><input aria-invalid=\\"false\\" name=\\"email\\" type=\\"text\\" class=\\"MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense\\" value=\\"\\"><fieldset aria-hidden=\\"true\\" class=\\"PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline\\"><legend class=\\"PrivateNotchedOutline-legendLabelled-3\\"><span>Email</span></legend></fieldset></div></div><div class=\\"MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense MuiInputLabel-outlined\\" data-shrink=\\"false\\">Password</label><div class=\\"MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense\\"><input aria-invalid=\\"false\\" name=\\"password\\" type=\\"password\\" class=\\"MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense\\" value=\\"\\"><fieldset aria-hidden=\\"true\\" class=\\"PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline\\"><legend class=\\"PrivateNotchedOutline-legendLabelled-3\\"><span>Password</span></legend></fieldset></div></div><div class=\\"MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense MuiInputLabel-outlined\\" data-shrink=\\"false\\">Age</label><div class=\\"MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense\\"><input aria-invalid=\\"false\\" name=\\"age\\" type=\\"number\\" class=\\"MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense\\" value=\\"\\"><fieldset aria-hidden=\\"true\\" class=\\"PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline\\"><legend class=\\"PrivateNotchedOutline-legendLabelled-3\\"><span>Age</span></legend></fieldset></div></div><button type=\\"submit\\">Validate</button></form>"`;
